hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmKeyInfo.java

```diff
@@ -789,6 +789,36 @@ public OmKeyInfo copyObject() {
     return builder.build();
   }
 
+  /**
+   * Return a new copy of the object.
+   */
+  @Override
+  public Object clone() throws CloneNotSupportedException {
+    OmKeyInfo omKeyInfo = (OmKeyInfo) super.clone();
+
+    omKeyInfo.metadata = new HashMap<>();
+    omKeyInfo.keyLocationVersions = new ArrayList<>();
+    omKeyInfo.acls = new ArrayList<>();
+
+    keyLocationVersions.stream().filter(keyLocationVersion ->
+            keyLocationVersion != null).forEach(keyLocationVersion ->
+            omKeyInfo.keyLocationVersions.add(
+                    new OmKeyLocationInfoGroup(keyLocationVersion.getVersion(),
+                            keyLocationVersion.getLocationList(),
```

Comment:
Should we deep copy the contents of this list as well? `OmKeyLocationInfo` extends `BlockLocationInfo` which is not immutable.

Reply:
I had the same question but looks like it is a deep copy 
```  /**
   * Use this expensive method only when absolutely needed!
   * It creates a new list so it is not an O(1) operation.
   * Use getLocationLists() instead.
   * @return a list of OmKeyLocationInfo
   */
  public List<OmKeyLocationInfo> getLocationList() {
    return locationVersionMap.values().stream().flatMap(List::stream)
        .collect(Collectors.toList());
  }```

Reply:
Hi @errose28 thanks for the review.
We could have done shallow copy in HeapEntry#getStatus and modify just the two fields which is required, but still cache value having same reference is present in HeapEntry and in future if some other fields gets modified there is still possibility of corruption by not doing deep copy to the modified field.

OmKeyLocationInfo is also deep copied as pointed out by Ritesh.

Reply:
@errose28 agreed that we need something more efficient. The entirety of `OzoneListStatusHelper.java` and listing logic needs to be revisited. Listing in general needs to be reviewed given that 99% of the use case does not care about the key info, to begin with. But getting something off the cache and returning a mutable reference is dangerous and this fix is needed to avoid corrupting keys.

Reply:
https://issues.apache.org/jira/browse/HDDS-8422

Reply:
The follow up jira is fine with me, as long as we are aware that list keys is probably going to be slower than before until it is resolved.